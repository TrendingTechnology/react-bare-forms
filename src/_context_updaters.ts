// Context Updaters
// Methods to update & track the state of all inputs, validators & general form metadata
//
// For example:
// const context = {
//     // other props...
//     metadata: {
//          inputs: {
//              email: {
//                  isValid: false,
//                  messages: ['Cannot be empty'],
//                  value: '',
//                  isTouched: false,
//              },
//              username: {
//                  isValid: true,
//                  messages: [],
//                  value: 'joebloggs',
//                  isTouched: true,
//              }
//          }
//     }
// };
import {IValidation} from "./validators";


export const updateValidationMetadata = (context: any, update: any) =>
    (fieldName: string, fieldValue: any, validation: IValidation): void => {
        if(typeof context.metadata === "undefined") {
            return;
        }
        if(!(fieldName in context.metadata)) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
            // Should update context.metadata value
        } else if(context.metadata[fieldName].value != fieldValue) {
            update({
                ...context,
                metadata: {
                    ...context.metadata,
                    [fieldName]: {
                        ...validation,
                        value: fieldValue,
                        isTouched: true,
                    }
                },
            });
        }

    };


