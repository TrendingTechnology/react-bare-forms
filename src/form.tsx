import {default as React, useEffect, useState} from "react";
import {updateStateFromPassedInContext} from "./_handlers";
import {IValidation} from "./validators";
import {updateRadioGroupMetadata, updateValidationMetadata} from "./_context_updaters";
import {IRadioField} from "./form-elements";


export interface IRadioGroupChildren {
    name: string;
    isChecked: boolean;
    disabled: boolean; // TODO
}

/** @internal */
export interface _IFormMetadata {
    messages: Array<string>;
    isValid: boolean;
    isTouched: boolean;
    value: any;
}

/** @internal */
export type TypeMetadata = { [k: string]: _IFormMetadata};
export type TypeRadioGroup = {[k: string]: IRadioGroupChildren};

export interface IMetdadata {
    fieldGroups: TypeRadioGroup;
    inputs: TypeMetadata;
}

/**
 * @interface **IForm** Exported Form interface available to the caller. Contains all the properties required by
 * the Form *RBF* Form's component.
 */
export interface IForm extends React.FormHTMLAttributes<HTMLFormElement> {
    /** The passed in state from the parent component */
    state: any;
    /** If the parent component is a class component then the context must contain the parent's **this** keyword. */
    context?: any;
    /** If the form's state is not directly contained at parent's state root object, then formKey needs to represent the key. */
    formKey?: string;
    /** Default is false. If set to true then the form state will be displayed just below the form component in `code` tags. */
    debug?: boolean;
    /** Default is True. If set to false, the  validators will not be called until the form is submitted. */
    dynamic?: boolean;
    /** The passed through props to the Form component */
    children?: any;
    /** Default set to false. If set to true then it will now use bootstrap styling or any extra elements. */
    bare?: boolean;
}

/** @internal */
export interface IFormContext {
    bare?: boolean;
    debug?: boolean;
    dynamic?: boolean;
    formKey?: string;
    metadata: TypeMetadata;
    state: any;
    updateParentState?: (e: React.ChangeEvent<any>, name: string) => void;
    updateFieldValidation?: (fieldName: string, fieldValue: any, validation: IValidation) => void;
    updateFormGroupMetadata?: (fieldGroupKey: string, radioProps: Array<{ props: IRadioField}>) => void;
}


const providerContext: IFormContext = {
    bare: false,
    state: {},
    formKey: null,
    debug: false,
    dynamic: true,
    metadata: { inputs: ({} as any), fieldGroups: ({} as any)},
};

/** @internal */
export const FormContext = React.createContext(providerContext);
/** @internal */
export const FormProvider = FormContext.Provider;
/**
 * @var Access the form state within a React Conext Api Consumer scope
 * @example For example:
 * ```
 * <Form state={this.state} context={this}>
 *      <FormConsumer>
 *      {({state}) => {
 *          return <div>{state} is equal {myState}</div>;
 *      }}
 *    </FormConsumer>
 * </Form>
 * ```
 * */
export const FormConsumer = FormContext.Consumer;


export const handleSubmit = (e: React.SyntheticEvent) => {
    e.preventDefault();
};

/**
 *
 * @param props
 * @constructor
 */
export const Submit = (props: any) => {
    return <button type="submit">Submit</button>;
};


/**
 * The main Form component
 * @param props
 *
 * ```
 * // Minimal setup for a RBF's Form component
 *
 *  const myState = {
 *      username: '',
 *  }
 *
 *  <Form state={myState}></Form>
 * ```
 * @constructor
 */
export const Form = (props: IForm) => {
    const [parentState, setParentState] = useState(props.state);

    // If the parent component is a class component, then the state needs to be updated from the parent context
    if(props.context) {
        useEffect(() => {
            props.context.setState({
                ...parentState,
            });
        }, [parentState]);
    }


    const [context, updateContext] = useState(providerContext);

    const _providerContext: IFormContext = {
        bare: props.bare || context.bare,
        state: props.state,
        formKey: props.formKey,
        debug: props.debug || context.debug,
        dynamic: props.dynamic || context.dynamic,
        metadata: context.metadata,
        updateParentState: updateStateFromPassedInContext(parentState, setParentState),
        updateFieldValidation: updateValidationMetadata(context, updateContext),
        updateFormGroupMetadata: updateRadioGroupMetadata(context, updateContext),
    };

    return (
        <FormProvider value={_providerContext}>
            <form onSubmit={handleSubmit} {...props}>{props.children}</form>
        </FormProvider>
    );
};
