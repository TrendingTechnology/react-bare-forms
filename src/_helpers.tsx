import * as React from "react";
import {handleChange} from "./handlers";
import {FieldTypes} from "./form-elements";
import {IFormContext} from "./form";
import {IValidation} from "./validators";

/**
 *
 * @param fieldType
 * @param name
 * @param context
 */
export const selectTextField = (fieldType: FieldTypes, name: any, context: IFormContext) => {
    switch(fieldType) {
        case "text": {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "email": {
            return <input type="email" className="form-control" onChange={handleChange(name, context)} />;
        }
        case "password": {
            // TODO confirm password
            return <input type="password" className="form-control" onChange={handleChange(name, context)} />;
        }
        default: {
            return <input type="text" className="form-control" onChange={handleChange(name, context)} />;
        }
    }
};

/**
 *
 * @param validationResult
 * @param context
 */
export const shouldShowValidation = (validationResult: IValidation, context: IFormContext) => {
    return (!validationResult.isValid && context.dynamic) ||
        (!validationResult.isValid && context.isSubmitted);
};


/**
 *
 * @param update
 * @param current
 */
export function setFormData(update: Function, current: IFormContext) {
    return (name: any, value: any) => {
        update({
            ...current,
            state: {
                ...current.state,
                [current.formKey]: {
                    ...current.state[current.formKey],
                    [name]: value,
                }
            }
        });
    };
}
