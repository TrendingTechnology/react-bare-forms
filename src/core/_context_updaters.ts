// Internal functions - Context Updaters
// Methods to update & track the state of all inputs, validators & general form metadata
import {IValidation} from "../validators";
import {IRadioField} from "../elements";
import {ICheckBoxesMetadata, IFileMetaData, IFormContext, IInputFieldMetadata, TypeMetadataNames,} from "../form";
import {IFile} from "./_file";


/** @internal */
type TypeUpdateValidation<T> = (updateData: T) => IFormContext;

/** @internal */
function _updateValidationContext<T>(context: IFormContext, type: TypeMetadataNames): TypeUpdateValidation<T> {
    return function(updateData: T) {
        return {
           ...context,
           metadata: {
               ...context.metadata,
               [type]: {
                   ...context.metadata[type],
                   ...updateData,
               },
           },
       };
    }
}

/** @internal */
function _isMatch(name: string, file: IFile, context: IFormContext): boolean {
    // Theres a bug here that makes file update twice, the second time wipes our all the data ..
    let fileInMetadata = context.metadata.files[name].file;
    if(!fileInMetadata && file || fileInMetadata && !file) return true;
    if(fileInMetadata) {
        if(fileInMetadata.name !== file.name) {
            return true;
        }
        // In case the user has duplicate filenames in different directories
        // we also coerce against the lastModified values.
        // @ts-ignore
        if(fileInMetadata.lastModified !== file.lastModified) {
            return true;
        }
    }
    return false;
}

/** @internal */
export const updateValidationMetadata = (context: any, update: any) => {
    /**
     * @param match
     *      - For input fields this is the passed value.
     *      - For Files this is the ReactRef.
     * @param name
     *      - For inputs & file inputs this is the name passed by the caller
     */
    return (name: string, match: any, validations: Array<IValidation>, type: TypeMetadataNames): void => {
        // Metadata field update types
        const updateInput = _updateValidationContext<IInputFieldMetadata>(context, "inputs");
        const updateFiles = _updateValidationContext<IFileMetaData>(context, "files");
        const updateCheckboxes = _updateValidationContext<ICheckBoxesMetadata>(context, "checkboxes");

        switch(type) {
            case "inputs": {
                if (!(name in context.metadata.inputs) || context.metadata.inputs[name].value !== match) {
                    let inputs = {
                        ...context.metadata.inputs,
                        [name]: {
                            validation: validations,
                            value: match,
                            isTouched: !!match,
                        },
                    };
                    const updatedData = updateInput(inputs);
                    update(updatedData);
                }
                break;
            }
            case "files": {
                if(!(name in context.metadata.files) || _isMatch(name, match, context)) {
                    let files = {
                        ...context.metadata.files,
                        [name]: {
                            validations: validations,
                            isTouched: name in context.metadata.files,
                            file: match,
                            refName: name,
                        }
                    };
                    update(updateFiles(files));
                }
                break;
            }
            case "checkboxes": {
                if(!(name in context.metadata.checkboxes) || context.metadata.checkboxes[name].isChecked !== match) {
                    let checkboxes = {
                        ...context.metadata.checkboxes,
                        [name]: {
                            validations: validations,
                            isTouched: name in context.metadata.checkboxes,
                            isChecked: match,
                        }
                    };
                    update(updateCheckboxes(checkboxes));
                }
                break;
            }
            default: {
                return null;
            }
        }
    };
};

/** @internal */
function _addFieldGroupToMetadata(radioProps: Array<{ props: IRadioField}>) {
    let fieldGroups = {};
    for(let radioVal of radioProps) {
        fieldGroups = {
            ...fieldGroups,
            [radioVal.props.name]: {
                name: radioVal.props.name,
                isChecked: radioVal.props.checked,
                disabled: radioVal.props.disabled || false,
                validation: radioVal.props.validators
            }
        }
    }
    return fieldGroups;
}

/** @internal */
export function shouldUpdateRadioGroupContext(radioProps: Array<{ props: IRadioField}>, context: any, fieldGroupKey: string) {
    for(let radio of radioProps) {
        let radioContext = context.metadata.fieldGroups[fieldGroupKey][radio.props.name];
        let radioState = radio.props;
        if(radioContext.isChecked !== radioState.checked) {
            return true;
        }
    }
}

/** @internal */
// This function will be called nth (the length of radioProps) times
// but will only update the context once per radio field group
export function updateRadioGroupMetadata(context: any, update: any) {
    return (fieldGroupKey: string, radioProps: Array<{ props: IRadioField}>) => {
        if(!(fieldGroupKey in context.metadata.fieldGroups) ||
            fieldGroupKey in context.metadata.fieldGroups && shouldUpdateRadioGroupContext(radioProps, context, fieldGroupKey)) {
            let contextUpdates = {
                ...context,
                metadata: {
                    ...context.metadata,
                    fieldGroups: {
                        ...context.metadata.fieldGroups,
                        [fieldGroupKey]: _addFieldGroupToMetadata(radioProps),
                    }
                }
            };
            update(contextUpdates);
        }
    }
}
