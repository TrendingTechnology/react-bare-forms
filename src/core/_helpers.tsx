import * as React from "react";
import {ReactElement} from "react";
import {METADATA_NAMES, TypeFieldValueTypes, TypeFormMetadata} from "../form";
import {IValidation} from "../validators";
import {FIELD_NAMES} from "../elements";
import {AbstractMetadata} from "../classes/_AbstractMetadata";


/** @internal */
interface IFormElementValidators {
    validators: Array<(...args: Array<any>) => IValidation>;
    name: string;
    value: any;
    type: METADATA_NAMES;
    className?: string;
    /* If a field is part of a group e.g Radio buttons then this is the name of parent field **/
    readonly parent?: string;
    metadata: AbstractMetadata<TypeFormMetadata>;
}
/** @internal */
type TypeValidationElement = { results: Array<IValidation>, styles: string };

/** @internal */
function ValidationResults(props: TypeValidationElement): ReactElement<TypeValidationElement> {
    const { results, styles } = props;
    return (<>{results.map((result: IValidation) =>
        result.messages.map((msg: string, index: number) =>
            <div key={index} className={styles}>{msg}</div>
        ))}</>);
}

/** @internal */
export function mergeDefaultCssWithProps(defaultValue: string, cssProps: any, bare: boolean): string {
    let cssStr = "";
    if(!bare) {
        cssStr += `${defaultValue} `;
    }
    if (cssProps) {
        cssStr += `${cssProps}`;
    }
    return cssStr;
}

/** @internal */
export function getMetadataNameType(type: FIELD_NAMES): METADATA_NAMES {
    switch(type) {
        case FIELD_NAMES.TEXT: {
            return METADATA_NAMES.INPUTS;
        }
        case FIELD_NAMES.EMAIL: {
            return METADATA_NAMES.INPUTS;
        }
        case FIELD_NAMES.PASSWORD: {
            return METADATA_NAMES.INPUTS;
        }
        case FIELD_NAMES.TEXTAREA: {
            return METADATA_NAMES.INPUTS;
        }
        // case FIELD_NAMES.RADIO: {
        //     return METADATA_NAMES.RADIO_GROUPS;
        // }
        case FIELD_NAMES.CHECKBOX: {
            return METADATA_NAMES.CHECKBOXES;
        }
        case FIELD_NAMES.SELECT: {
            return METADATA_NAMES.INPUTS;
        }
        case FIELD_NAMES.FILE: {
            return METADATA_NAMES.FILES;
        }
        default: {
            return METADATA_NAMES.INPUTS;
        }
    }
}

/** @internal */
export function getFieldValueType(type: FIELD_NAMES): TypeFieldValueTypes {
    switch(type) {
        case FIELD_NAMES.TEXT: {
            return "value";
        }
        case FIELD_NAMES.EMAIL: {
            return "value";
        }
        case FIELD_NAMES.PASSWORD: {
            return "value";
        }
        case FIELD_NAMES.TEXTAREA: {
            return "value";
        }
        case FIELD_NAMES.CHECKBOX: {
            return "checked";
        }
        case FIELD_NAMES.SELECT: {
            return "value";
        }
        case FIELD_NAMES.FILE: {
            return "file";
        }
        default: {
            return "value";
        }
    }
}
