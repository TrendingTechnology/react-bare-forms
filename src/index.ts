import {Form, FormConsumer} from "./form";
import {
    CheckBoxField,
    EmailField, FileField,
    PasswordField, RadioField, RadioGroup,
    SelectField,
    SubmitButton,
    TextAreaField,
    TextInputField
} from "./elements";
import {areFieldsEqual, isChecked, isEmailValid, isFieldEmpty, isFile, isRadioChecked} from "./validators";
import {createFileRef, getFileFromRef} from "./uncrontrolled";




export {
    // Uncontrolled
    createFileRef,
    getFileFromRef,
    // Validators
    areFieldsEqual,
    isChecked,
    isEmailValid,
    isFieldEmpty,
    isFile,
    isRadioChecked,
    // Form Components
    FormConsumer,
    Form,
    SubmitButton,
    TextInputField,
    EmailField,
    PasswordField,
    TextAreaField,
    CheckBoxField,
    SelectField,
    FileField,
    RadioField,
    RadioGroup,
}
