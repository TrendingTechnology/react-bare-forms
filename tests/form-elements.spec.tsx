import * as React from "react";
import renderer from "react-test-renderer";
import {
    EmailField, FormGroup,
    PasswordField, TextAreaField,
    TextInputField,
} from "../src/form-elements";
import {FormConsumer, FormProvider} from "../src/form";
import {isEmailValid, isFieldEmpty} from "../src/validators";
import {updateValidationMetadata} from "../src/_context_updaters";
import {FormElementValidators} from "../src/_helpers";


describe("#TextInputField()", () => {
    it("should render an input field with the bootstrap styled tags", () => {
        let state = {
            username: "joe",
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextInputField name="username" value={state.username} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a single input tag with no bootstrap styles", () => {
        let state = {
            username: "",
        };
        let contextDefault = {
            bare: true,
            state: {},
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };
        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextInputField name="username" value={state.username} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });


    it("it should pass in props", () => {
        let state = {
            username: "joebloggs"
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let testFormRederer: any = renderer.create(
            <FormProvider value={contextDefault}>
                <TextInputField name="username" value={state.username} />
            </FormProvider>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props.value).toEqual(state.username);

    });

    it("it should display validation errors", () => {
        let state = {
            username: "a",
        };

        let _contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, (t: any) => {})
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextInputField
                    name="username"
                    value={state.username}
                    validators={[isFieldEmpty(5)]}
                />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("it should pass validation props", () => {
        let state = {
            username: "a",
        };

        let _contextDefault = {
            bare: true,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };


        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, updateMetadata)
        };

        function updateMetadata(t: any) {
            contextDefault["metadata"] = t.metadata;
        }

        let testFormRederer: any = renderer.create(
            <div>
            <FormProvider value={contextDefault}>
                <TextInputField
                    name="username"
                    value={state.username}
                    validators={[isFieldEmpty(5)]}
                />
            </FormProvider>
            </div>
        );

        const testFormInstance = testFormRederer.root;
        // expect(testFormInstance.props.value).toEqual(state.username);
        expect(testFormInstance
            .find(FormElementValidators)
            .props
            .children
            .props
            .value
            .metadata
        ).toEqual({"username": {
            isTouched: true,
            isValid: false,
            messages: [
              "Must be at least 5 characters",
            ],
            value: "a",
             }
        });

    });

});

describe("#EmailField()", () => {
    it("should render an email field with the bootstrap styled tags", () => {
        let state = {
            email: "joe@joe.com",
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <EmailField name="email" value={state.email} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a single input tag with no bootstrap styles", () => {
        let state = {
            email: "email@email.com",
        };
        let contextDefault = {
            bare: true,
            state: {},
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };
        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <EmailField name="email" value={state.email} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });


    it("it should pass in props", () => {
        let state = {
            email: "joe@joe.com"
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let testFormRederer: any = renderer.create(
            <FormProvider value={contextDefault}>
                <EmailField name="email" value={state.email} />
            </FormProvider>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props.value).toEqual(state.email);

    });

    it("it should display validation errors", () => {
        let state = {
            email: "joejoe",
        };

        let _contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, (t: any) => {})
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <EmailField
                    name="email"
                    value={state.email}
                    validators={[isEmailValid()]}
                />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("it should pass validation props", () => {
        let state = {
            email: "a",
        };

        let _contextDefault = {
            bare: true,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };


        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, updateMetadata)
        };

        function updateMetadata(t: any) {
            contextDefault["metadata"] = t.metadata;
        }

        let testFormRederer: any = renderer.create(
            <div>
                <FormProvider value={contextDefault}>
                    <EmailField
                        name="email"
                        value={state.email}
                        validators={[isEmailValid()]}
                    />
                </FormProvider>
            </div>
        );

        const testFormInstance = testFormRederer.root;
        // expect(testFormInstance.props.value).toEqual(state.username);
        expect(testFormInstance
            .find(FormElementValidators)
            .props
            .children
            .props
            .value
            .metadata
        ).toEqual({"email": {
                isTouched: true,
                isValid: false,
                messages: [
                    "Must be a valid email",
                ],
                value: "a",
            }
        });

    });
});

describe("#PasswordField()", () => {
    it("should render an input field with the bootstrap styled tags", () => {
        let state = {
            password: "joe",
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <PasswordField name="password" value={state.password} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a single input tag with no bootstrap styles", () => {
        let state = {
            password: "",
        };
        let contextDefault = {
            bare: true,
            state: {},
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };
        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <PasswordField name="password" value={state.password} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });


    it("it should pass in props", () => {
        let state = {
            password: "joebloggs"
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let testFormRederer: any = renderer.create(
            <FormProvider value={contextDefault}>
                <PasswordField name="password" value={state.password} />
            </FormProvider>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props.value).toEqual(state.password);

    });

    it("it should display validation errors", () => {
        let state = {
            password: "a",
        };

        let _contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, (t: any) => {})
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <PasswordField
                    name="password"
                    value={state.password}
                    validators={[isFieldEmpty(5)]}
                />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("it should pass validation props", () => {
        let state = {
            password: "a",
        };

        let _contextDefault = {
            bare: true,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };


        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, updateMetadata)
        };

        function updateMetadata(t: any) {
            contextDefault["metadata"] = t.metadata;
        }

        let testFormRederer: any = renderer.create(
            <div>
                <FormProvider value={contextDefault}>
                    <PasswordField
                        name="password"
                        value={state.password}
                        validators={[isFieldEmpty(5)]}
                    />
                </FormProvider>
            </div>
        );

        const testFormInstance = testFormRederer.root;
        // expect(testFormInstance.props.value).toEqual(state.password);
        expect(testFormInstance
            .find(FormElementValidators)
            .props
            .children
            .props
            .value
            .metadata
        ).toEqual({"password": {
                isTouched: true,
                isValid: false,
                messages: [
                    "Must be at least 5 characters",
                ],
                value: "a",
            }
        });

    });
});

describe("#<TextAreaField />", () => {
    it("should render an input field with the bootstrap styled tags", () => {
        let state = {
            about_text: "joe",
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextAreaField name="about_text" value={state.about_text} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a single input tag with no bootstrap styles", () => {
        let state = {
            about_text: "",
        };
        let contextDefault = {
            bare: true,
            state: {},
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };
        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextAreaField name="about_text" value={state.about_text} />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });


    it("it should pass in props", () => {
        let state = {
            about_text: "joebloggs"
        };
        let contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let testFormRederer: any = renderer.create(
            <FormProvider value={contextDefault}>
                <TextAreaField name="about_text" value={state.about_text} />
            </FormProvider>
        );

        const testFormInstance = testFormRederer.root;
        expect(testFormInstance.props.value).toEqual(state.about_text);

    });

    it("it should display validation errors", () => {
        let state = {
            about_text: "a",
        };

        let _contextDefault = {
            bare: false,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };

        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, (t: any) => {})
        };

        let component = renderer.create(
            <FormProvider value={contextDefault}>
                <TextAreaField
                    name="about_text"
                    value={state.about_text}
                    validators={[isFieldEmpty(5)]}
                />
            </FormProvider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("it should pass validation props", () => {
        let state = {
            about_text: "a",
        };

        let _contextDefault = {
            bare: true,
            state: state,
            formKey: "",
            debug: false,
            dynamic: true,
            metadata: {},
        };


        let contextDefault = {
            ..._contextDefault,
            updateFieldValidation: updateValidationMetadata(_contextDefault, updateMetadata)
        };

        function updateMetadata(t: any) {
            contextDefault["metadata"] = t.metadata;
        }

        let testFormRederer: any = renderer.create(
            <div>
                <FormProvider value={contextDefault}>
                    <TextAreaField
                        name="about_text"
                        value={state.about_text}
                        validators={[isFieldEmpty(5)]}
                    />
                </FormProvider>
            </div>
        );

        const testFormInstance = testFormRederer.root;
        // expect(testFormInstance.props.value).toEqual(state.about_text);
        expect(testFormInstance
            .find(FormElementValidators)
            .props
            .children
            .props
            .value
            .metadata
        ).toEqual({"about_text": {
                isTouched: true,
                isValid: false,
                messages: [
                    "Must be at least 5 characters",
                ],
                value: "a",
            }
        });

    });
});

