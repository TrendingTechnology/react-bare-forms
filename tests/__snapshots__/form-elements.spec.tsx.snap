// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#<TextAreaField /> it should display validation errors 1`] = `
<div
  className="form-group"
>
  <label />
  <textarea
    className="form-control "
    name="about_text"
    onChange={[Function]}
    rows={5}
  />
  <div
    className="alert mt-2 alert-danger undefined"
  >
    Must be at least 5 characters
  </div>
</div>
`;

exports[`#<TextAreaField /> should render a single input tag with no bootstrap styles 1`] = `
<textarea
  className=""
  name="about_text"
  onChange={[Function]}
  rows={5}
/>
`;

exports[`#<TextAreaField /> should render an input field with the bootstrap styled tags 1`] = `
<div
  className="form-group"
>
  <label />
  <textarea
    className="form-control "
    name="about_text"
    onChange={[Function]}
    rows={5}
  />
</div>
`;

exports[`#EmailField() it should display validation errors 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="email"
    onChange={[Function]}
    type="email"
  />
  <div
    className="alert mt-2 alert-danger undefined"
  >
    Must be a valid email
  </div>
</div>
`;

exports[`#EmailField() should render a single input tag with no bootstrap styles 1`] = `
<input
  className=""
  name="email"
  onChange={[Function]}
  type="email"
/>
`;

exports[`#EmailField() should render an email field with the bootstrap styled tags 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="email"
    onChange={[Function]}
    type="email"
  />
</div>
`;

exports[`#PasswordField() it should display validation errors 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="password"
    onChange={[Function]}
    type="password"
  />
  <div
    className="alert mt-2 alert-danger undefined"
  >
    Must be at least 5 characters
  </div>
</div>
`;

exports[`#PasswordField() should render a single input tag with no bootstrap styles 1`] = `
<input
  className=""
  name="password"
  onChange={[Function]}
  type="password"
/>
`;

exports[`#PasswordField() should render an input field with the bootstrap styled tags 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="password"
    onChange={[Function]}
    type="password"
  />
</div>
`;

exports[`#TextInputField() it should display validation errors 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="username"
    onChange={[Function]}
    type="text"
  />
  <div
    className="alert mt-2 alert-danger undefined"
  >
    Must be at least 5 characters
  </div>
</div>
`;

exports[`#TextInputField() should render a single input tag with no bootstrap styles 1`] = `
<input
  className=""
  name="username"
  onChange={[Function]}
  type="text"
/>
`;

exports[`#TextInputField() should render an input field with the bootstrap styled tags 1`] = `
<div
  className="form-group"
>
  <input
    className="form-control "
    name="username"
    onChange={[Function]}
    type="text"
  />
</div>
`;
